packages/langium/src/documents/document-builder.ts

```diff
@@ -54,55 +50,72 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
-            get diagnostics() {
-                if (!diagnostics) {
-                    diagnostics = validator.validateDocument(document);
-                }
-                return diagnostics;
-            }
+            diagnostics: await this.validate(document, cancelToken, true)
         };
     }
 
-    validate(document: LangiumDocument): Diagnostic[] | undefined {
-        let diagnostics: Diagnostic[] | undefined;
+    protected async validate(document: LangiumDocument, cancelToken = CancellationToken.None, forceDiagnostics = false): Promise<Diagnostic[]> {
+        let diagnostics: Diagnostic[] = [];
         const validator = this.documentValidator;
-        if (this.connection) {
-            diagnostics = validator.validateDocument(document);
-            // Send the computed diagnostics to VS Code.
-            this.connection.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });
+        if (this.connection || forceDiagnostics) {
+            diagnostics = await validator.validateDocument(document, cancelToken);
+            if (this.connection && !cancelToken.isCancellationRequested) {
+                // Send the computed diagnostics to VS Code.
+                this.connection.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });
+            }
+        }
+        if (!cancelToken.isCancellationRequested) {
+            document.state = DocumentState.Validated;
         }
         return diagnostics;
     }
 
-    documentChanged(uri: URI): void {
+    async documentChanged(uri: URI, cancelToken = CancellationToken.None): Promise<void> {
         this.langiumDocuments.invalidateDocument(uri);
         const newDocument = this.langiumDocuments.getOrCreateDocument(uri);
-        this.indexManager.update(newDocument);
-        this.langiumDocuments.all.filter(doc => this.isAffected(doc, uri)).forEach(
-            doc => {
-                this.build(doc);
-                this.indexManager.update(doc);
+        await delayNextTickAfter();
+        if (cancelToken.isCancellationRequested) {
+            return;
+        }
+        await this.buildDocument(newDocument, cancelToken);
+    }
+
+    protected async buildDocument(document: LangiumDocument, cancelToken: CancellationToken): Promise<void> {
+        const affectedDocuments = this.indexManager.getEffectedDocuments(document);
+        affectedDocuments.forEach(e => e.reset());
+        const relevantDocuments = Array.from(new Set([
+            document,
+            ...affectedDocuments,
+            // Also include all documents haven't completed the document lifecycle yet
+            ...this.langiumDocuments.all.filter(e => e.state < DocumentState.Validated)
+        ]));
+        await this.runCancelable(relevantDocuments, DocumentState.Processed, cancelToken, doc => this.process(doc));
+        await this.runCancelable(relevantDocuments, DocumentState.Indexed, cancelToken, doc => this.indexManager.update(doc, cancelToken));
```

Comment:
I think we need to switch these two phases: first index, then process, because processing may involve resolving cross-references. Indexing has the constraint that no cross-references must be resolved in that phase (the same is true in Xtext).

The main reason for this is that I would propose to include precomputation of types for languages with typed expressions in the processing phase. Type computation involves a lot of following cross-references.